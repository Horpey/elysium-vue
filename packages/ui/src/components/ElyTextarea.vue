<script lang="ts" setup>
import type { TextareaHTMLAttributes } from 'vue'
import type { ElyInputSizeKey } from '~/types/ElyInput'

interface Props extends /* @vue-ignore */ TextareaHTMLAttributes {
  name?: string
  modelValue?: string | number
  required?: boolean
  placeholder?: string
  disabled?: boolean
  inputClass?: string
  inputId?: string
  rows?: number
  size?: ElyInputSizeKey
  resize?: boolean
  onInput?: (event: Event) => void
  onBlur?: (event: Event) => void
  onChange?: (event: Event) => void
}

const props = withDefaults(defineProps<Props>(), {
  name: undefined,
  modelValue: undefined,
  required: false,
  placeholder: undefined,
  disabled: false,
  inputClass: undefined,
  inputId: undefined,
  rows: 3,
  resize: true,
  size: 'MEDIUM',
  onInput: () => { },
  onBlur: () => { },
  onChange: () => { },
})

const emit = defineEmits(['update:modelValue'])

const classes = computed(() => {
  const baseClass = 'border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:opacity-75'

  let classString = baseClass
  let sizeClass = ''

  switch (props.size) {
    case 'SMALL':
      sizeClass = 'text-xs px-2 py-1'
      break
    case 'MEDIUM':
      sizeClass = 'text-sm px-3 py-2'
      break
    case 'LARGE':
      sizeClass = 'text-base px-4 py-3'
      break
  }

  if (props.class)
    classString += ` ${props.class}`

  if (props.inputClass)
    classString += ` ${props.inputClass}`

  if (sizeClass)
    classString += ` ${sizeClass}`

  if (props.resize)
    classString += ' resize'
  else classString += ' resize-none'

  return classString
})

function onInput(event: Event) {
  props.onInput(event)
  emit('update:modelValue', (event.target as HTMLInputElement).value)
}

function onBlur(event: Event) {
  props.onBlur(event)
}

function onChange(event: Event) {
  props.onChange(event)
}
</script>

<template>
  <span class="relative">
    <textarea
      :id="inputId"
      :value="modelValue"
      :name="name"
      :rows="rows"
      :required="required"
      :disabled="disabled"
      :placeholder="placeholder"
      v-bind="$attrs"
      :class="classes"
      @input="onInput"
      @blur="onBlur"
      @change="onChange"
    />
    <slot />
  </span>
</template>
