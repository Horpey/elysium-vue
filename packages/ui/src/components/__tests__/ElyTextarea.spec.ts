import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import ElyTextarea from '../ElyTextarea.vue'

describe('elyTextarea', () => {
  it('renders with default props', async () => {
    const wrapper = mount(ElyTextarea)

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert default values
    expect(wrapper.find('textarea').exists()).toBe(true)
    expect(wrapper.find('textarea').attributes('rows')).toBe('3')
    expect(wrapper.find('textarea').classes()).toContain('border-koromiko-600')
    expect(wrapper.find('textarea').classes()).toContain('resize')

    // Assert slot content is not provided
    expect(wrapper.text()).toBe('')

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom props and triggers input event', async () => {
    const wrapper = mount(ElyTextarea, {
      props: {
        name: 'description',
        modelValue: 'This is a description.',
        required: true,
        placeholder: 'Enter your description',
        disabled: false,
        inputClass: 'custom-input-class',
        inputId: 'custom-input-id',
        rows: 5,
        size: 'LARGE',
        resize: false,
        onInput: vi.fn(),
        onBlur: vi.fn(),
        onChange: vi.fn(),
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert custom values
    expect(wrapper.find('textarea').attributes('name')).toBe('description')
    expect(wrapper.find('textarea').attributes('id')).toBe('custom-input-id')
    expect(wrapper.find('textarea').attributes('rows')).toBe('5')
    expect(wrapper.find('textarea').attributes('placeholder')).toBe('Enter your description')
    expect(wrapper.find('textarea').classes()).toContain('custom-input-class')
    expect(wrapper.find('textarea').classes()).toContain('resize-none')
    // ... add more assertions based on your styling and custom props

    // Trigger input event
    await wrapper.find('textarea').setValue('Updated description.')

    // Assert that the onInput function was called
    expect(wrapper.props().onInput).toHaveBeenCalled()

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })
})
