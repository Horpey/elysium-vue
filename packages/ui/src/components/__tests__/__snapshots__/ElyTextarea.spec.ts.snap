// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elyTextarea > renders with custom props and triggers input event 1`] = `"<span class="relative"><textarea id="custom-input-id" name="description" rows="5" required="" placeholder="Enter your description" class="border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:opacity-75 custom-input-class text-base px-4 py-3 resize-none"></textarea></span>"`;

exports[`elyTextarea > renders with default props 1`] = `"<span class="relative"><textarea rows="3" class="border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:opacity-75 text-sm px-3 py-2 resize"></textarea></span>"`;
